crates/neon/src/lifecycle.rs

```diff
@@ -54,6 +59,176 @@ pub(crate) struct InstanceData {
 
     /// Shared `Channel` that is cloned to be returned by the `cx.channel()` method
     shared_channel: Channel,
+
+    /// Table of user-defined global cells.
+    globals: GlobalTable,
+}
+
+#[derive(Default)]
+pub(crate) struct GlobalTable {
+    cells: Vec<GlobalCell>,
+}
+
+pub(crate) type GlobalCellValue = Box<dyn Any + Send + 'static>;
+
+pub(crate) enum GlobalCell {
+    /// Uninitialized state.
+    Uninit,
+    /// Intermediate "dirty" state representing the middle of a `get_or_try_init` transaction.
+    Trying,
+    /// Fully initialized state.
+    Init(GlobalCellValue),
+}
+
+impl GlobalCell {
+    /// Establish the initial state at the beginning of the initialization protocol.
+    /// This method ensures that re-entrant initialization always panics (i.e. when
+    /// an existing `get_or_try_init` is in progress).
+    fn pre_init<F>(&mut self, f: F)
+    where
+        F: FnOnce() -> GlobalCell,
+    {
+        match self {
+            GlobalCell::Uninit => {
+                *self = f();
+            }
+            GlobalCell::Trying => panic!("attempt to reinitialize Global during initialization"),
+            GlobalCell::Init(_) => {}
+        }
+    }
+
+    pub(crate) fn get<'cx, 'a, C>(cx: &'a mut C, id: usize) -> Option<&mut GlobalCellValue>
```

Comment:
I haven't tested removing it, but it looks like this doesn't need to be `pub(crate)`.

Reply:
It's used by `crate::instance::Global::get()`.